### PR TITLE
refact(auth): cache login role in auth client to avoid sending rpc repeatedly

### PR DESCRIPTION
**Pull request by imbajin**
TODO: also need be compatible with token way (consider put 'usrname + passwd' together into caches's key)

PS: remove useless cache in `HugegraphAuthProxy`

before cache:
image

after cache in auth client:
image

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java [hunk 1]
```diff
@@ -95,7 +101,12 @@ public void setup(HugeConfig config) {
         String graphName = config.get(ServerOptions.AUTH_GRAPH_STORE);
         String graphPath = config.getMap(ServerOptions.GRAPHS).get(graphName);
         E.checkArgument(graphPath != null,
-                        "Invalid graph name '%s'", graphName);
+                        "Can't find graph name '%s' in config '%s' at " +
+                        "'rest-server.properties' to store auth information, " +
+                        "please ensure the value of '%s' matches it correctly",
+                        graphName, ServerOptions.GRAPHS,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java:107]: ServerOptions.GRAPHS.name()?

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java [hunk 2]
```diff
@@ -125,8 +138,19 @@ public RolePermission authenticate(String username, String password) {
...
 
-        RolePermission role = this.graph().authManager().loginUser(username,
-                                                                   password);
+        // TODO: Be compatible with JWT(token) way
+        Id usrname = IdGenerator.of(username);
+        UserRoleCache userRole = this.usersRoleCache.getOrFetch(usrname, r -> {
+            RolePermission rolePermission = this.graph().authManager()
+                                                .loginUser(username, password);
+            return new UserRoleCache(rolePermission, password);
+        });
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java:147]: prefer to move to hugegraph auth proxy

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java (hunk 1 above)
Comment (imbajin) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java:107]: GRAPHS.name() only shows `graphs` in error (not clear enough)

GRAPHS.toString() returns name + value (seems more friendly for user to find out mismatch)

Comment (javeme) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java:107]: ok

File: .github/workflows/stale.yml [hunk 3]
```diff
@@ -27,7 +27,7 @@ jobs:
         days-before-pr-stale: 30
         days-before-pr-close: 180
         operations-per-run: 25
-        start-date: '2020-12-01T00:00:00Z'
+        start-date: '2018-10-01T00:00:00Z'
```
Comment (javeme) [inline .github/workflows/stale.yml:30]: can commit separately

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java [hunk 4]
```diff
@@ -625,6 +626,15 @@ public static synchronized CoreOptions instance() {
                     10
             );
 
+    public static final ConfigOption<Double> AUTH_AUDIT_LOG_RATE =
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/CoreOptions.java:629]: move to AuthOptions

File: .github/workflows/stale.yml (hunk 3 above)
Comment (imbajin) [reply .github/workflows/stale.yml:30]: OK, separate it into `github-ci` pr later

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeAuthenticator.java [hunk 5]
```diff
@@ -167,6 +171,10 @@ public String username() {
             return this.getName();
         }
 
+        public Id userId() {
+            return this.id;
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeAuthenticator.java:175]: rename userId

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java [hunk 6]
```diff
@@ -114,7 +121,15 @@ public HugeGraphAuthProxy(HugeGraph hugegraph) {
         this.hugegraph = hugegraph;
         this.taskScheduler = new TaskSchedulerProxy(hugegraph.taskScheduler());
         this.authManager = new AuthManagerProxy(hugegraph.authManager());
+        this.auditLimiters = this.cache("audit-log-limiter", -1L);
+        this.usersRoleCache = this.cache("users-role",
+                                         Duration.ofMinutes(10L).toMillis());
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java:126]: add option

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java [hunk 7]
```diff
@@ -1371,8 +1400,12 @@ public RolePermission rolePermission(AuthElement element) {
         public UserWithRole validateUser(String username, String password) {
             // Can't verifyPermission() here, validate first with tmp permission
             Context context = setContext(Context.admin());
+
             try {
-                return this.authManager.validateUser(username, password);
+                Id mixedUserId = IdGenerator.of(username + password);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java:1405]: userKey

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java [hunk 8]
```diff
@@ -1386,8 +1419,11 @@ public UserWithRole validateUser(String username, String password) {
         public UserWithRole validateUser(String token) {
             // Can't verifyPermission() here, validate first with tmp permission
             Context context = setContext(Context.admin());
+
             try {
-                return this.authManager.validateUser(token);
+                return usersRoleCache.getOrFetch(IdGenerator.of(token), id -> {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeGraphAuthProxy.java:1424]: userKey = IdGenerator.of(token)

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java [hunk 9]
```diff
@@ -46,28 +46,33 @@ private HugeGraph graph() {
...
+        if (requireInitAdminUser(authManager)) {
             HugeUser admin = new HugeUser(HugeAuthenticator.USER_ADMIN);
             admin.password(StringEncoding.hashPassword(password));
             admin.creator(HugeAuthenticator.USER_SYSTEM);
             authManager.createUser(admin);
         }
     }
 
+    private boolean requireInitAdminUser(AuthManager authManager) {
+        return StandardAuthManager.isLocal(authManager) &&
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java:72]: seems can get authManager = this.graph().hugegraph().authManager();

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java (hunk 9 above)
Comment (imbajin) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/auth/StandardAuthenticator.java:72]: shall we use this `this.authManager()` directly?
```java
    @Override
    public AuthManager authManager() {
        return this.graph().authManager();
    }
```

File: hugegraph-core/src/main/java/com/baidu/hugegraph/config/AuthOptions.java [hunk 10]
```diff
@@ -92,4 +94,22 @@ public static synchronized AuthOptions instance() {
...
+                    "The max rate of audit log output per user, " +
+                    "default value is 1000 records per second.",
+                    rangeDouble(0.0, Double.MAX_VALUE),
+                    1000.0
+            );
+
+    public static final ConfigOption<Long> AUTH_USER_ROLE_CACHE_EXPIRE =
+            new ConfigOption<>(
+                    "auth.user_role_cache_expire",
+                    "The expire time in minutes of user role cache " +
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/config/AuthOptions.java:110]: change the unit to ms to keep the same style

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeAuthenticator.java [hunk 11]
```diff
@@ -153,12 +155,14 @@ public static HugeAuthenticator loadAuthenticator(HugeConfig conf) {
...
 
         private final RolePermission role;
+        private final Id userId;
         private String client; // peer
 
         public User(String username, RolePermission role) {
             super(username);
             E.checkNotNull(username, "username");
             E.checkNotNull(role, "role");
+            this.userId = IdGenerator.of(username);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/HugeAuthenticator.java:165]: can pass from UserWithRole?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java [hunk 12]
```diff
@@ -472,7 +475,7 @@ public UserWithRole validateUser(String username, String password) {
 
     @Override
     public UserWithRole validateUser(String token) {
-        String username = (String) this.tokenCache.get(IdGenerator.of(token));
+        String username = this.tokenCache.get(IdGenerator.of(token));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java:478]: replace CACHE_EXPIRE with tokenCache.expire() at line 493.
and improve the way to invalid pwd and token cache.

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java [hunk 13]
```diff
@@ -48,12 +48,16 @@
 
 public class StandardAuthManager implements AuthManager {
 
-    private static final long CACHE_EXPIRE = Duration.ofDays(1L).toMillis();
+    private static final long TOKEN_EXPIRE = Duration.ofDays(1L).toMillis();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java:51]: remove it

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java [hunk 14]
```diff
@@ -453,7 +470,7 @@ public String loginUser(String username, String password)
                                                  username,
                                                  AuthConstant.TOKEN_USER_ID,
                                                  user.id.asString());
-        String token = this.tokenGenerator.create(payload, CACHE_EXPIRE);
+        String token = this.tokenGenerator.create(payload, TOKEN_EXPIRE);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java:473]: also add a option

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java [hunk 15]
```diff
@@ -48,7 +48,7 @@
 
 public class StandardAuthManager implements AuthManager {
 
-    private static final long TOKEN_EXPIRE = Duration.ofDays(1L).toMillis();
+    private static long TOKEN_EXPIRE;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java:51]: remove static

File: hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java [hunk 16]
```diff
@@ -48,7 +48,7 @@
 
 public class StandardAuthManager implements AuthManager {
 
-    private static long TOKEN_EXPIRE;
+    private long tokenExpire;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/auth/StandardAuthManager.java:51]: set final and move to line 69

**Review by javeme — APPROVED** [review]

**Review by Linary — APPROVED** [review]

**Pull request merged**